File: ios/ZenIdModule.m
//
//  ZenIdModule.m
//  ZenIdNative
//
//  Created by Josef Kvapil on 15.05.2024.
//

#import <Foundation/Foundation.h>
#import "React/RCTBridgeModule.h"

@interface RCT_EXTERN_MODULE(ZenIdModule, NSObject)

RCT_EXTERN_METHOD(initializeSdk:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(isAuthorized:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(authorize:(NSString *)token resolve:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(getChallengeToken:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(selectProfile:(NSString *)profile resolve:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(activateTakeNextDocumentPicture:(nonnull NSNumber *)viewTag resolve:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

@end

File: ios/ZenIdModule.swift
import Foundation
import RecogLib_iOS

@objc(ZenIdModule)
class ZenIdModule: NSObject, RCTBridgeModule {
  weak var bridge: RCTBridge?
  static func moduleName() -> String! {
    return "ZenIdModule"
  }

 static func requiresMainQueueSetup() -> Bool {
    return true
  }
 

  @objc
  func isAuthorized(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
    resolve(true)
  }

  @objc
  func getChallengeToken(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
    resolve(ZenidSecurity.getChallengeToken())
  }

  @objc
  func authorize(_ token: String, resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
    let authorized = ZenidSecurity.authorize(responseToken: token)
    if authorized {
      resolve("Authorized")
    } else {
      reject("AuthorizationError", "Failed to authorize", nil)
    }
  }

  @objc
  func initializeSdk(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    resolve("SDK initialized successfully")
  }

  @objc
  func selectProfile(_ profile: String, resolve: RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    let profileSelected = ZenidSecurity.selectProfile(name: profile)
    if profileSelected {
      resolve("Profile selected")
    } else
    {
      reject("ProfileSelectionError", "Failed to select profile", nil)
    }
  }

  @objc
   func activateTakeNextDocumentPicture(_ viewTag: NSNumber, resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
     DispatchQueue.main.async {
       if let view = self.bridge?.uiManager.view(forReactTag: viewTag) as? DocumentPictureView {
         view.activateNextDocumentPicture()
         resolve(nil)
       } else {
         reject("ViewNotFound", "Could not find view with tag \(viewTag)", nil)
       }
     }
   }
  
  
}
